### PR TITLE
Fix Security Vulnerabilities

### LINKED ISSUE #86: (XSS) Cross-Site Scripting Vulnerability - /view_resources/
## Summary
The **Note** parameter of a resource is vulnerable to cross-site scripting.
The user input data is not properly sanitized in the resource creation process as well as when the resource is viewed via a shared link.

## Vulnerability Details
**Severity**: High
**OWASP Category**: A3 – Cross-Site Scripting (XSS)
**OWASP Page**: https://www.owasp.org/index.php/Cross-site_Scripting_(XSS)

## Proof of Concept URL
1. Navigate to this URL https://threerstore.herokuapp.com/view_resources/WzgsInNlY3JldCJd.DuaiXw.zg0dyJwNWEs63kYgpvW0OW0mUfs
2. Notice the pop-up alert box indicating that the XSS payload was executed successfully

## Remediation
User input should be validated and escaped properly during the resource creation process as well as during its reflection back to the user under the `/view_resources` view.

The following chunk of code is where user input should be escaped and sanitized properly before it's shown back to the user.
https://github.com/TsimpDim/3RStore/blob/49ccb10f49cc7aee6d825aa28484ea9a9475f11e/_3RStore/templates/resources_public.html#L67-L79

And, the following code is where user input data should be escaped before it is inserted into the database to prevent arbitrary code injection.
https://github.com/TsimpDim/3RStore/blob/49ccb10f49cc7aee6d825aa28484ea9a9475f11e/_3RStore/views.py#L307-L313

## Note
Currently the app has crashed due to another critical bug. Once it's up and running again I will update this issue to include more detailed steps on how to exactly reproduce the issue.

### LINKED ISSUE #87: Invalid Tag input data crashes the app and causes Denial of Service
## Reproduction Steps

1.  Navigate to https://threerstore.herokuapp.com/login and authenticate with your account credentials
2. Then, under https://threerstore.herokuapp.com/resources create a new resource
3. On the resource's tag, enter the following payload, `test">`
4. Create the resource and notice that the application has crashed.

## Note
As the application has currently crashed, this issue will be updated with more details once it's back up and running.

### PR DESCRIPTION
**Pull request by iamaldi**
Hello @TsimpDim . This pull request resolves #86 and also resolves #87 as well as makes sure that foreign key constraints are not a headache anymore when a user tries to delete his (own) account (hopefully 😃 )

### CODE REVIEW

**Comment by iamaldi** [comment]
The second commit (392560e1d7fac169a9b1a77f885601f0cbcc6810)  of this pull request resolves #89

**Comment by iamaldi** [comment]
Commit 9792ad3dcd638ec8b665f64f7b95b3d34c9c2256 introduces some new functions to reduce the code footprint of the application as well as to introduce some new security features to prevent unauthorized access to other user's data.

Specifically the following changes are made by merging this commit:
User is redirected to the login page if he/she performs an action that needs the user to be logged in.
If a user is already logged out then calling `/logout` sends a message telling the user is already logged out without needing to clear sessions etc.
If a user is already logged in, the application redirects the user to the `/resources` page notifying the user with a `Already logged in` message
User can only delete his/her own resources
a `.` has been added to every `flash()` message as to make them more appealing and eye catching.

**Comment by iamaldi** [comment]
Improve overall security and add necessary common HTTP Security Headers in order to prevent malicious activity. Commit b22472e3a49df350a9a3d6332cd7692441f5a039 tries to and partially resolves #90

File: _3RStore/views.py [hunk 1]
```diff
@@ -15,6 +15,28 @@
 from anytree import Node, RenderTree, find, AsciiStyle, NodeMixin, AnyNode, PreOrderIter
 import re as r
 
+# Checks if the provided input consists of alphanumerical characters only (include a comma)
+def inputValidation(userInput):
```
Comment (TsimpDim) [inline _3RStore/views.py]: Use snake_case for function names

File: _3RStore/views.py [hunk 2]
```diff
@@ -15,6 +15,28 @@
...
+    # Returns False if the user input is invalid
+    # Returns True if the user input is valid
+    if r.match("^[A-Za-z0-9_, -]*$", userInput):
+        return True
+    else:
+        flash('Invalid characters were inserted.', 'danger')
+        return False
+
+# Checks if a user is logged in
+def isUserLoggedIn():
```
Comment (TsimpDim) [inline _3RStore/views.py]: Use snake_case for function names

File: _3RStore/views.py [hunk 3]
```diff
@@ -15,6 +15,28 @@
...
+    if 'logged_in' in session and session.get('logged_in') == True:
+        return True
+    else:
+        return False
```
Comment (TsimpDim) [inline _3RStore/views.py]: This is a one liner

Comment (TsimpDim) [reply _3RStore/views.py]: return 'logged_in' in session and session.get('logged_in') == True

File: _3RStore/views.py [hunk 4]
```diff
@@ -15,6 +15,28 @@
...
+
+# Redirects user to the login page
+def redirectToLoginPage():
```
Comment (TsimpDim) [inline _3RStore/views.py]: Use snake_case for function names

File: _3RStore/views.py (hunk 1 above)
Comment (TsimpDim) [inline _3RStore/views.py]: Add those helper functions to a separate file

File: _3RStore/views.py [hunk 5]
```diff
@@ -48,107 +70,130 @@ def register():
 
             cur.close()
             conn.commit()
-
-        flash('You are now registered', 'success')
-        return redirect(url_for('login'))
+            flash('You are now registered.', 'success')
```
Comment (TsimpDim) [inline _3RStore/views.py]: Weird indentation?

File: _3RStore/views.py [hunk 6]
```diff
@@ -48,107 +70,130 @@ def register():
...
+            return redirect(url_for('login'))
 
     return render_template('register.html', form=form)
 
 # User Login
 @app.route('/login', methods=['GET', 'POST'])
 def login():
+    # Check if user is not logged in
+    if isUserLoggedIn() == False:
```
Comment (TsimpDim) [inline _3RStore/views.py]: Not pythonic. Use the 'not' operator.

File: _3RStore/views.py [hunk 7]
```diff
@@ -48,107 +70,130 @@ def register():
...
-
-    return render_template('login.html')
 
+            cur.close()
+        else:
+            return render_template('login.html')
+    else:
+        # Otherwise redirect to the `resources` page
+        flash('You are already logged in.', 'success')
+        return redirect(url_for('resources'))
```
Comment (TsimpDim) [inline _3RStore/views.py]: Set a newline for code readability

File: _3RStore/views.py [hunk 8]
```diff
@@ -48,107 +70,130 @@ def register():
...
+        flash('You are already logged in.', 'success')
+        return redirect(url_for('resources'))    
 # Logout
 @app.route('/logout')
 def logout():
-    session.clear()
-    flash('You are now logged out', 'success')
-    return redirect('/')
+    # Check if user is not logged out
+    if isUserLoggedIn() == True:
```
Comment (TsimpDim) [inline _3RStore/views.py]: Not pythonic -- redundant `== True`

File: _3RStore/views.py [hunk 9]
```diff
@@ -48,107 +70,130 @@ def register():
...
+    else:
+        flash('You are already logged out.', 'success')
+        return redirect('/')
 
 # Delete account
 @app.route('/delacc', methods=['POST'])
 def delacc():
-    user_id = request.form.get('user_id')
+    # Check if user is logged in
+    if isUserLoggedIn() == True:
```
Comment (TsimpDim) [inline _3RStore/views.py]: Not pythonic -- redundant == True

File: _3RStore/views.py [hunk 10]
```diff
@@ -48,107 +70,130 @@ def register():
...
+        # Extract user_id from the current session
+        user_id = session['user_id']
+    
+        cur = conn.cursor()
+        try:
+            # First delete from `resources` so as not to violate foreign key constraints
+            cur.execute('DELETE FROM resources WHERE user_id = %s',
+            ([user_id])
```
Comment (TsimpDim) [inline _3RStore/views.py]: Unnecessary deviation from using single element tuples

File: _3RStore/views.py [hunk 11]
```diff
@@ -48,107 +70,130 @@ def register():
...
 
-    cur = conn.cursor()
-    try:
-        # First delete from `resources` so as not to violate foreign key constraints
-        cur.execute('DELETE FROM resources WHERE user_id = %s',
-        (user_id,)
-        )
+            # Delete from `trash` as not to violate foreign key constraints
+            cur.execute('DELETE FROM trash WHERE user_id = %s',
+            ([user_id])
```
Comment (TsimpDim) [inline _3RStore/views.py]: Unnecessary deviation from using single element tuples

File: _3RStore/views.py [hunk 12]
```diff
@@ -48,107 +70,130 @@ def register():
...
+            )
 
-        cur.execute('DELETE FROM users WHERE id = %s',
-        (user_id,)
-        )
-    except DatabaseError:
-        cur.rollback()
+            # Finally, remove the user from `users`
+            cur.execute('DELETE FROM users WHERE id = %s',
+            ([user_id])
```
Comment (TsimpDim) [inline _3RStore/views.py]: Unnecessary deviation from using single element tuples

File: _3RStore/views.py [hunk 13]
```diff
@@ -168,15 +213,15 @@ def chpass():
         cur.close()
         conn.commit()
 
-        flash('Password changed successfully', 'success')
+        flash('Password changed successfully.', 'success')
```
Comment (TsimpDim) [inline _3RStore/views.py]: Unnecessary change, but sure, why not

File: _3RStore/views.py [hunk 14]
```diff
@@ -168,15 +213,15 @@ def chpass():
...
         return redirect(url_for('options'))
     return render_template('chng_password.html', form=form)
 
 # Options
 @app.route('/options')
 def options():
 
-    if session.get('logged_in'):
+    if isUserLoggedIn() == True:
```
Comment (TsimpDim) [inline _3RStore/views.py]: Not pythonic -- redundant `== True`

File: _3RStore/views.py [hunk 15]
```diff
@@ -38,7 +60,7 @@ def register():
             time()).strftime('%Y-%m-%d %H:%M:%S')
 
         if not conn:
-            flash('Could not connect to database', 'error')
+            flash('Could not connect to database.', 'error')
```
Comment (TsimpDim) [inline _3RStore/views.py]: Unnecessary change, but sure, why not

File: _3RStore/views.py [hunk 16]
```diff
@@ -200,7 +245,7 @@ def options():
         
         return render_template('options.html', sort=sort, criteria=criteria, tags=all_tags, view=view)
     else:
-        flash('You must be logged in to access the options page', 'warning')
+        flash('You must be logged in to access the options page.', 'warning')
```
Comment (TsimpDim) [inline _3RStore/views.py]: Unnecessary change, but sure, why not

File: _3RStore/views.py [hunk 17]
```diff
@@ -328,7 +374,7 @@ def add_resource():
         cur.close()
         conn.commit()
 
-        flash('Resource created successfully', 'success')
+        flash('Resource created successfully.', 'success')
```
Comment (TsimpDim) [inline _3RStore/views.py]: Trailing whitespace

File: _3RStore/views.py [hunk 18]
```diff
@@ -534,63 +586,64 @@ def edit_res(user_id, re_id):
...
+        flash('All resources were deleted successfully.', 'danger')
+        return redirect(url_for('resources'))
+    else:
+        return redirectToLoginPage()
 
 # Filtered delete
 @app.route("/fildel", methods=['POST'])
 def fildel():
+    # Check if user is logged in
+    if isUserLoggedIn() == True:
```
Comment (TsimpDim) [inline _3RStore/views.py]: Not pythonic -- redundant `== True`

File: _3RStore/views.py [hunk 19]
```diff
@@ -610,7 +663,7 @@ def remtag():
         cur.close()
         conn.commit()
 
-    flash('Tag(s) removed successfully', 'danger')
+    flash('Tag(s) removed successfully.', 'danger')
```
Comment (TsimpDim) [inline _3RStore/views.py]: Full stops are a really unnecessary change

File: _3RStore/views.py (hunk 3 above)
Comment (TsimpDim) [reply _3RStore/views.py]: Perhaps even `return 'logged_in' in session and session.get('logged_in')`

File: _3RStore/__init__.py [hunk 20]
```diff
@@ -1,14 +1,44 @@
 import json
 import os
-from flask import Flask
+from flask import Flask, request
 from flask_mail import Mail
+from flask_cors import CORS
```
Comment (TsimpDim) [inline _3RStore/__init__.py]: Just noticed, you didn't update requirements.txt

File: _3RStore/views.py [hunk 21]
```diff
@@ -515,13 +562,18 @@ def edit_res(user_id, re_id):
...
-            note = form.note.data.replace('\n','</br>') # Save newlines as </br> to display them properly later
-            tags = form.tags.data
+            link = Markup.escape(form.link.data)
+            note = Markup.escape(form.note.data.replace('\n','</br>')) # Save newlines as </br> to display them properly later
+            tags = Markup.escape(form.tags.data)
 
             # If not empty format for proper insertion into postgresql
             if tags:
-                tags = '{' + str(tags).lower() + '}'
+                if inputValidation(tags):
```
Comment (TsimpDim) [inline _3RStore/views.py]: Validation happens at form level - check out `forms.py`

File: _3RStore/views.py [hunk 22]
```diff
@@ -305,13 +350,14 @@ def add_resource():
 
     form = forms.ResourceForm(request.form)
     if request.method == 'POST' and form.validate():
-        title = form.title.data
-        link = urllib.parse.unquote(form.link.data)
-        note = form.note.data.replace('\n','</br>') # So we can show the newlines in the note section
+        # Escape user input using Markup
+        title = Markup.escape(form.title.data)
+        link = urllib.parse.unquote(Markup.escape(form.link.data))
+        note = Markup.escape(form.note.data.replace('\n','</br>')) # So we can show the newlines in the note section
```
Comment (TsimpDim) [inline _3RStore/views.py]: Newlines are rendered as `</br>` instead of actual new lines

**Comment by TsimpDim** [comment]
#90, #89, #87 and #86  all seem be fixed. 

----

Changes requested:

1. Update *requirements.txt* (Current *requirements.txt* was build with [Pigar](https://pypi.org/project/pigar/))

2. For consistency use single element tuples `(x,)` instead of lists `[x]` when passing arguments to psycopg2 queries

3. The helper functions (`inputValidation`,`isUserLoggedIn` and `redirectToLoginPage`) should not be in the views file (`views.py`) but in a separate one

    3.1.  Input validation for forms happens at `forms.py` -> Insert your `inputValidation` code there
    3.2.  Function `redirectToLoginPage` is only used twice. I'd say either make all redirects a function or remove this one altogether

4. Function and variable names should use snake_case ([PEP8 section](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)) 

5. Condense the already mentioned pieces of code as the Python syntax allows

6. Fix newlines on notes no longer being allowed

--- 



### !! Don't forget to pull the changes made after your pull request

**Comment by TsimpDim** [comment]
Closed due to inactivity. Changes will be fixed and posted under a different pull request & branch at a later time.

**Pull request closed without merging**